js/treetest.js

```diff
@@ -115,8 +130,80 @@ var tasks = {
 			var copyOfHistory = $.extend(true, [], this.answers[this.idx]);
 			expandToNode(copyOfHistory);
 		}
-	}
+		this.onSet.forEach(callback => callback(number));
+	},
+	onSet: [],
+	onSubmit: [],
 }
+
+var socket = {
+	_socket: null,
+
+	connect: function() {
+		this._socket = io();
+		this._socket.on('connect', () => {
+			console.debug('Socket connected');
+			this._socket.on('disconnect', () => {
+				console.debug('Socket disconnected');
+			})
+		})
+	},
+
+	emitPageLoadEvent: function() {
+		this._emitEvent({}, 'page load');
+	},
+
+	emitActivateNodeEvent: function(node) {
+		const data = {
+			node: $('#tree').jstree().get_path(node),
+			current_task_index: tasks.idx,
+		};
+		this._emitEvent(data, 'activate node');
+	},
+
+	emitOpenNodeEvent: function(node) {
+		const data = {
+			node: $('#tree').jstree().get_path(node),
+			current_task_index: tasks.idx,
+		};
+		this._emitEvent(data, 'open node');
+	},
+
+	emitCloseNodeEvent: function(node) {
```

Comment:
There is a setting `Hide Siblings` in tree test which I think automatically collapses open nodes. Will that cause an event capture as well, not sure if we would want that. 

I could not test it since things are not working for me

Reply:
Yes. Given `Hide siblings` is enabled, if you open one node, and this automatically causes another node to close, that means we get a 'close node' event.

It's already the case that you can sometimes get more than one event emitted per user action. This wouldn't be the only such case. For example, if you open a parent node by activating it, you get both an 'open node' event and an 'activate node' event.

@rvango doesn't seem to like this behaviour either, I think he's working on explaining why that is and what he would like to happen instead

Reply:
To keep it simple, I would have like to have `open` and `activated` to be the same event, and possibly be renamed as `selected`.

I might do that in a follow-up ticket because I reckon that might be a bit more tricky using the generic `check_callback()` that gets called for any events

```js
$('#your_tree_id').jstree({
    'core': {
        'check_callback': function(operation, node, node_parent, node_position, more) { 
          ...
    }
```